### ISSUE TITLE
optimizer.step() and lr_scheduler.step()

### ISSUE THREAD

**Issue by Bryce1010** [ildoonet/pytorch-gradual-warmup-lr#8]
UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate

**Comment by ildoonet** [comment]
@Bryce1010 Thanks. I will update thie repo for torch 1.4

**Comment by ildoonet** [comment]
@Bryce1010 I check the warning, but it seems to be unavoidable.

Unlike most of the schedulers, which start with the given initial learning rate and adapt it step by step, this gradual warmup scheduler should modify the initial learning rate as zero before any backward update on weights. 

But if you want to avoid this warning message, there is a walk-around. See the latest code. I call 'optimizer.step()' with zero gradients, right after I create schedulers.

**Comment by lucasbrynte** [comment]
I think this work-around does the job (meaning nothing, except suppressing the warning), but I also think perhaps the design of the scheduler was based on a wrong assumption that the .step() method is only called directly by the user. On the contrary, PyTorch indeed makes an initial call to .step() from the scheduler's constructor, via the ._initial_step() function (at least in recent versions, perhaps ever since version 1.1). Before this implicit initial step, `self._last_epoch = -1`, and after the initial step, `self._last_epoch = 0`. Consequently, while I do not fully understand why .get_lr() alone could not compute the correct learning rates (without any preceding .step() calls), .step() will in any case have been called once in the scheduler constructor.

Perhaps by simply replacing most occurences of `self._last_epoch` by `self._last_epoch+1` in the current implementation, one could preserve the functionality even if .step() is only called at the end of each epoch by the user..?

I do however think the whole scheduler would be easier / less error-prone to implement using the built-in PyTorch LR scheduler [LinearLR](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LinearLR.html) for the warmup part, optionally chained with one or more other schedulers (the equivalent of "after_scheduler") using [SequentialLR](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.SequentialLR.html).
